pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java

```diff
@@ -332,6 +340,31 @@ private void validateRequest(BrokerRequest brokerRequest) {
             "Value for 'LIMIT' (" + limit + ") exceeds maximum allowed value of " + _queryResponseLimit);
       }
     }
+
+    // Checks whether the query contains non-existent columns.
+    // Table name has already been verified before hitting this line.
+    if (requestParams.isValidateQuery()) {
+      String tableName = brokerRequest.getQuerySource().getTableName();
+      Schema schema = _tableSchemaCache.getIfTableSchemaPresent(tableName);
+      if (schema != null) {
+        RequestInfo requestInfo = RequestUtils.preComputeRequestInfo(brokerRequest);
+        // gets all the columns from broker request
+        Set<String> columnsFromBrokerRequest = requestInfo.getAllColumns();
+        // filters out virtual columns in the query.
+        columnsFromBrokerRequest.removeIf(column -> column.startsWith("$"));
+        columnsFromBrokerRequest.removeAll(schema.getColumnNames());
+        if (!columnsFromBrokerRequest.isEmpty()) {
+          _brokerMetrics.addMeteredTableValue(tableName, BrokerMeter.QUERY_NON_EXISTENT_COLUMNS, 1L);
+          throw new RuntimeException(
```

Comment:
Can we throw a more specific exception so that we can throw 400 bad request instead of a 500 internal server error?

Reply:
This runtime exception will be captured in caller's exception block and thus throws a `QUERY_VALIDATION_EXCEPTIONS` back to the client. So we're good here.